docs/doxygen/man/man3/Simulation.3
.TH "Simulation" 3 "Version 1.0.0" "Warehouse Simulator" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Simulation \- The \fBSimulation\fP class manages the overall department store simulation\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fR#include <Simulation\&.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBSimulation\fP ()"
.br
.RI "Construct a new \fBSimulation\fP object\&. "
.ti -1c
.RI "\fBstatus\fP \fBconductCycle\fP ()"
.br
.RI "Conducts a simulation cycle, processing events for the current cycle\&. "
.ti -1c
.RI "\fBstatus\fP \fBrespondToEvent\fP (\fBEvent\fP &event)"
.br
.RI "Responds to a specific event\&. "
.ti -1c
.RI "void \fBrun\fP ()"
.br
.RI "Initiates the simulation by running the main loop\&. "
.ti -1c
.RI "\fBstatus\fP \fBprocessEvents\fP ()"
.br
.RI "Processes all scheduled events\&. "
.ti -1c
.RI "QString \fBgenerateReport\fP ()"
.br
.RI "Generates a report summarizing simulation results\&. "
.in -1c
.SH "Detailed Description"
.PP 
The \fBSimulation\fP class manages the overall department store simulation\&. 

This class orchestrates the simulation process, including event handling, warehouse management, product operations, and reporting\&. 
.PP
Definition at line \fB28\fP of file \fBSimulation\&.h\fP\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Simulation::Simulation ()"

.PP
Construct a new \fBSimulation\fP object\&. Sets up the initial state of the simulation, including loading settings and preparing the environment\&. 
.PP
Definition at line \fB11\fP of file \fBSimulation\&.cpp\fP\&.
.SH "Member Function Documentation"
.PP 
.SS "\fBstatus\fP Simulation::conductCycle ()"

.PP
Conducts a simulation cycle, processing events for the current cycle\&. Processes events and updates the simulation state for each cycle\&. 
.PP
\fBReturns\fP
.RS 4
Status of the cycle (success or failure)\&. 
.RE
.PP

.PP
Definition at line \fB191\fP of file \fBSimulation\&.cpp\fP\&.
.SS "QString Simulation::generateReport ()"

.PP
Generates a report summarizing simulation results\&. Compiles data from the simulation into a formatted report string\&. 
.PP
\fBReturns\fP
.RS 4
QString A summary report of the simulation results\&. 
.RE
.PP

.PP
Definition at line \fB332\fP of file \fBSimulation\&.cpp\fP\&.
.SS "\fBstatus\fP Simulation::processEvents ()"

.PP
Processes all scheduled events\&. Invokes conductCycle to process events for the current cycle\&. 
.PP
\fBReturns\fP
.RS 4
Status of proccessing events)\&. 
.RE
.PP

.PP
Definition at line \fB178\fP of file \fBSimulation\&.cpp\fP\&.
.SS "\fBstatus\fP Simulation::respondToEvent (\fBEvent\fP & event)"

.PP
Responds to a specific event\&. Handles the given event based on its type and updates the simulation state accordingly\&. 
.PP
\fBParameters\fP
.RS 4
\fIevent\fP The event to handle\&. 
.RE
.PP
\fBReturns\fP
.RS 4
Status of event handling (success or failure)\&. 
.RE
.PP

.PP
Definition at line \fB229\fP of file \fBSimulation\&.cpp\fP\&.
.SS "void Simulation::run ()"

.PP
Initiates the simulation by running the main loop\&. Starts the simulation loop, generating cycles, processing events, and generating reports\&. 
.PP
Definition at line \fB110\fP of file \fBSimulation\&.cpp\fP\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for Warehouse Simulator from the source code\&.


src/Simulation/Simulation.cpp
/**
 * @file Simulation.cpp
 * @brief Source file for the Simulation class.
 */

#include "Simulation.h"

/**
 * @brief Construct a new Simulation object.
 */
Simulation::Simulation()
{
    std::cout << "Preparing simulation" << std::endl;

    currentTime = QDateTime::currentDateTime();

    QFile settings("settings.csv");

    QFile::rename("SimulationReport.csv", "SimulationReportOld.csv");
    QFile::remove("SimulationReport.csv");

    if(!settings.open(QIODevice::ReadOnly | QIODevice::Text))
    {
        std::cerr << "Error: Can't open settings file." << std::endl;
        return;
    }
    QTextStream in(&settings);

    seed = 100;
    int productIdCounter = 1;

    while(!in.atEnd())
    {
        QString line = in.readLine();
        QStringList fields = line.split(',');

        if(fields.isEmpty())
        {
            continue;
        }

        if(fields[0] == "Warehouse")
        {
            productIdCounter = 1;
            if(fields.size() < 3)
            {
                std::cerr << "Error: Incomplete warehouse data." << std::endl;
                continue;
            }
            std::cout << "Adding warehouse." << std::endl;
            QString location = fields[1];
            double capacity = fields[2].toDouble();
            Warehouses.append(Warehouse(location, capacity));
        }
        else if(fields[0] == "Product")
        {
            if(fields.size() < 4)
            {
                std::cerr << "Error: Incomplete product data." << std::endl;
                continue;
            }
            std::cout << "Adding product: ";
            Warehouse& selectWarehouse = Warehouses.last();
            QString name = fields[1];
            double price = fields[2].toDouble();
            int quantity = fields[3].toInt();
            int productId = productIdCounter++;

            if(selectWarehouse.addProduct(name, price, quantity, productId) == SUCCESS)
            {
                std::cout << "SUCCESS" << std::endl;
            }
            else
            {
                std::cerr << "ERROR: Unable to add product." << std::endl;
            }
        }
        else if(fields[0] == "Cycles")
        {
            if(fields.size() < 5)
            {
                std::cerr << "Error: Incomplete cycle data." << std::endl;
                continue;
            }
            std::cout << "Setting currentCycle: ";
            int currentCycle = fields[4].toInt();
            this -> currentCycle = currentCycle;
            std::cout << currentCycle << std::endl;
        }
        else if(fields[0] == "Seed")
        {
            if(fields.size() < 6)
            {
                std::cerr << "Error: Incomplete seed data." << std::endl;
                continue;
            }
            std::cout << "Setting seed: ";
            int readSeed = fields[5].toInt();
            this -> seed = QRandomGenerator::global() -> generate()/readSeed;
            std::cout << readSeed << std::endl;
        }
    }

    settings.close();
}

/**
 * @brief Initiates the simulation by running the main loop.
 */
void Simulation::run()
{
    std::cout << "Running simulation with " << currentCycle << " cycles." << std::endl;

    if(Warehouses.isEmpty())
    {
        std::cout << "Incorrect settings" << std::endl;
        exit(1);
    }

    std::cout << "Generating cycles." << std::endl;

    int cycles = currentCycle;

    while(currentCycle > 0)
    {
        std::cout << "Processing cycle " << cycles - currentCycle << "." << std::endl;
        int minProductsAvailable = std::numeric_limits<int>::max();
        int totalProducts = 0;

        for (Warehouse& warehouse : Warehouses)
        {
            const QList<Product>& productList = warehouse.getProductList();
            if (productList.isEmpty())
            {
                std::cout << "Warning: No products available in warehouse at " << warehouse.getLocation().toStdString() << std::endl;
                continue;
            }

            int numberOfEvents = QRandomGenerator::global()->bounded(productList.size()) + 1;

            std::cout << "Number of sale events to be generated for warehouse at " << warehouse.getLocation().toStdString() << ": " << numberOfEvents << std::endl;

            for (int event = 0; event < numberOfEvents; ++event)
            {
                events.append(Event::generateEvent("Sell product", seed));
                std::cout << "\033[33mInfo: Generating event - Sell product\033[0m" << std::endl;
            }
        }

        if (events.isEmpty())
        {
            std::cout << "No events generated in this cycle." << std::endl;
        }
        else
        {
            QList<Warehouse> restoreWarehouse = Warehouses;

            if (processEvents() == ERROR && currentCycle != cycles)
            {
                Warehouses = restoreWarehouse;
            }

            QString cycleReport = generateReport();
            std::cout << cycleReport.toStdString() << std::endl;
            events.clear();
        }

        std::cout << "Cycle " << cycles - currentCycle << " completed." << std::endl;
        currentCycle--;
    }
    std::cout << "Simulation completed." << std::endl;
}


/**
 * @brief Processes all scheduled events.
 */
status Simulation::processEvents()
{
    if(conductCycle() == ERROR)
    {
        std::cout << "Error while processing cycle" << std::endl;
        return ERROR;
    }
    return SUCCESS;
}

/**
 * @brief Conducts a simulation cycle, processing events for the current cycle.
 */
status Simulation::conductCycle()
{
    qint64 deltaTime = 0;

    int successEvents = 0;

    for(Event& event : events)
    {
        if(respondToEvent(event) == ERROR)
        {
            std::cout << "Error while processing event." << std::endl;
        }
        else
        {
            successEvents++;
        }

        QDateTime Time = event.getTime();

        if(Time.msecsTo(currentTime) > deltaTime)
        {
            deltaTime = Time.msecsTo(currentTime);
        }
    }

    if(successEvents == 0)
    {
        return ERROR;
    }

    events.clear();
    currentTime = currentTime.addSecs(deltaTime);
    return SUCCESS;
}

/**
 * @brief Responds to a specific event.
 */
status Simulation::respondToEvent(Event& event)
{
    int warehouseId = QRandomGenerator::global() -> bounded(Warehouses.size());
    Warehouse& warehouse = Warehouses[warehouseId];

    if(event.getEventType() == "Sell product")
    {
        try
        {
            const QList<Product>& productList = warehouse.getProductList();
            if (productList.isEmpty())
            {
                std::cerr << "ERROR: No products to sell." << std::endl;
                return ERROR;
            }

            int productId = QRandomGenerator::global()->bounded(productList.size()) + productList.begin()->productId;
            std::cout << "Attempting to sell" << productId;
            status result = warehouse.sell(1, productId);

            if (result == ERROR)
            {
                std::cout << "ERROR: Unable to sell product. Product doesn't exists." << std::endl;
                return ERROR;
            }
        }
        catch (const std::exception& e)
        {
            std::cerr << "General Exception caught: " << e.what() << std::endl;
            return ERROR;
        }
    }
    else if (event.getEventType() == "Add product")
    {
        try
        {
            const QList<Product>& productList = warehouse.getProductList();
            int productId = QRandomGenerator::global()->bounded(productList.size());
            const Product& product = productList[productId];

            status result;

            while(warehouse.checkStatus() != FULLY)
            {
                result = warehouse.changeQuantity(product.getQuantity() + 1, productId);
            }

            if (result == ERROR)
            {
                std::cerr << "ERROR: Unable to add product quantity." << std::endl;
                return ERROR;
            }
        }
        catch (const std::exception& e)
        {
            std::cerr << "General Exception caught: " << e.what() << std::endl;
            return ERROR;
        }
    }
    else if (event.getEventType() == "Transfer product")
    {
        try
        {
            const QList<Product>& productList = warehouse.getProductList();
            int productId = QRandomGenerator::global()->bounded(productList.size());
            const Product& productToTransfer = productList[productId];

            auto targetWarehouseIt = Warehouses.end();
            for (auto it = Warehouses.begin(); it != Warehouses.end(); ++it)
            {
                if (it->getLocation() != warehouse.getLocation() && it-> checkStatus() == FULLY)
                {
                    targetWarehouseIt = it;
                    break;
                }
            }

            if (targetWarehouseIt != Warehouses.end())
            {
                Warehouse& targetWarehouse = *targetWarehouseIt;
                int transferQuantity = productToTransfer.getQuantity();
                warehouse.changeQuantity(warehouse.getQuantity(productId) - transferQuantity, productId);
                targetWarehouse.changeQuantity(targetWarehouse.getQuantity(productId) + transferQuantity, productId);
            }
            else
            {
                std::cerr << "ERROR: No target warehouse found for product transfer." << std::endl;
                return ERROR;
            }
        }
        catch (const std::exception& e)
        {
            std::cerr << "General Exception caught: " << e.what() << std::endl;
            return ERROR;
        }
    }

    return SUCCESS;
}

/**
 * @brief Generates a report summarizing simulation results.
 */
QString Simulation::generateReport()
{
    QString csvReport;
    int id = 0;

    std::cout << "Generating report" << std::endl;

    for(Warehouse& warehouse : Warehouses)
    {
        QList<Report::ProductReport> productNames;
        double featureOperationalCosts = 0;
        for(const Product& product : warehouse.getProductList())
        {
            Report::ProductReport productReport;
            productReport.name = product.getName();
            productReport.price = product.getPrice();
            productReport.quantity = product.getQuantity();

            featureOperationalCosts += productReport.price + productReport.quantity;

            productNames.append(productReport);
        }

        static int salesId = 0;

        WarehouseReport warehouseReport(id++, warehouse.getCurrentCapacity(), productNames, Report::getOperationalCosts(), Report::getNetProfit());
        SalesReport salesReport(salesId++, currentTime, productNames, Report::getOperationalCosts(), Report::getNetProfit());

        Report::setOperationalCosts(Report::getOperationalCosts()+featureOperationalCosts);

        csvReport.append(warehouseReport.generateReport());
        csvReport.append(salesReport.generateReport());
    }

    // Save the CSV report to a file
    QFile csvFile("SimulationReport.csv");

    if(csvFile.open(QIODevice::Append | QIODevice::Text))
    {
        QTextStream out(&csvFile);
        out << "\n";
        out << csvReport;
        csvFile.close();
    }
    else
    {
        std::cerr << "Error while trying to write the CSV report to file." << std::endl;
    }

    return csvReport;
}
